src/filters/matches.rs

```diff
@@ -0,0 +1,159 @@
+            metrics_registry,
+        }
+    }
+}
+
+/// The actual behaviour of the matches filter, the only thing that changes
+/// between read and write is which config and direction is called.
+macro_rules! matches_filter {
+    ($this:ident, $context:ident, $direction:ident, $config:ident) => {{
+        let ctx = $context;
+        let config = match $this.config.$config.as_ref() {
+            Some(value) => value,
+            None => return Some(ctx.into()),
+        };
+
+        for branch in &config.branches {
+            if ctx.metadata.get(&config.metadata_key) == Some(&branch.value) {
+                let args = CreateFilterArgs::fixed(
+                    $this.filter_registry.clone(),
+                    $this.metrics_registry.clone(),
+                    branch.config.as_ref(),
+                )
+                .with_metrics_registry($this.metrics_registry.clone());
+
+                let instance = $this.filter_registry.get(&branch.filter, args).ok()?;
+
+                return instance.filter.$direction(ctx);
+            }
+        }
+
+        None
+    }};
+}
+
+impl Filter for Matches {
+    #[cfg_attr(feature = "instrument", tracing::instrument(skip(self, ctx)))]
+    fn read(&self, ctx: ReadContext) -> Option<ReadResponse> {
+        matches_filter!(self, ctx, read, on_read)
+    }
+
+    #[cfg_attr(feature = "instrument", tracing::instrument(skip(self, ctx)))]
+    fn write(&self, ctx: WriteContext) -> Option<WriteResponse> {
+        matches_filter!(self, ctx, write, on_write)
+    }
+}
+
+struct MatchesFactory;
+
+impl MatchesFactory {
+    pub fn new() -> Self {
+        MatchesFactory
+    }
+}
+
+impl FilterFactory for MatchesFactory {
+    fn name(&self) -> &'static str {
+        NAME
+    }
+
+    fn create_filter(&self, args: CreateFilterArgs) -> Result<FilterInstance, Error> {
+        let (config_json, config) = self
+            .require_config(args.config)?
+            .deserialize::<Config, ProtoConfig>(self.name())?;
+
+        let filter = Matches::new(config, args.filter_registry.clone(), args.metrics_registry);
+        Ok(FilterInstance::new(
+            config_json,
+            Box::new(filter) as Box<dyn Filter>,
+        ))
+    }
+}
+
+#[derive(Debug, serde::Deserialize, serde::Serialize, PartialEq)]
+#[serde(deny_unknown_fields)]
+struct Config {
+    on_read: Option<DirectionalConfig>,
+    on_write: Option<DirectionalConfig>,
+}
+
+impl TryFrom<ProtoConfig> for Config {
+    type Error = ConvertProtoConfigError;
+
+    fn try_from(_value: ProtoConfig) -> Result<Self, Self::Error> {
+        todo!()
+    }
+}
+
+#[derive(Debug, serde::Deserialize, serde::Serialize, PartialEq)]
+struct DirectionalConfig {
+    metadata_key: String,
+    branches: Vec<Branch>,
+}
+
+#[derive(Debug, serde::Deserialize, serde::Serialize, PartialEq)]
+struct Branch {
+    value: crate::metadata::Value,
+    filter: String,
+    config: Option<serde_yaml::Value>,
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
I'd like for us to add some tests either here or as integration tests etc that exercises more functionality of the filter, like that we call read/write, we skip branches and select the correct branch when they're multiple branches involved and we drop a packet or forward it when there's no match based on the config.